docarray/array/mixins/find.py

```diff
@@ -0,0 +1,268 @@
+from typing import overload, Optional, Union, Dict, Tuple, Callable, TYPE_CHECKING
+from ...math.helper import top_k, minmax_normalize, update_rows_x_mat_best
+from ...score import NamedScore
+import numpy as np
+
+
+if TYPE_CHECKING:
+    from ...types import T, ArrayType
+
+    from ... import Document, DocumentArray
+
+
+class FindMixin:
+    """ A mixin that provides find functionality to DocumentArrays  """
+
+    @overload
+    def find(self: 'T', query: 'ArrayType', **kwargs):
+        ...
+
+    @overload
+    def find(self: 'T', query: Union['Document', 'DocumentArray'], **kwargs):
+        ...
+
+    @overload
+    def find(self: 'T', query: Dict, **kwargs):
+        ...
+
+    @overload
+    def find(self: 'T', query: str, **kwargs):
+        ...
+
+    def find(
+        self: 'T',
+        query: Union['DocumentArray', 'Document', 'ArrayType', Dict, str],
+        **kwargs,
+    ) -> 'DocumentArray':
+        from ...math.ndarray import to_numpy_array, get_array_type
+        from ... import Document, DocumentArray
+
+        if isinstance(query, (dict, str)):
+            return self.filter(query, **kwargs)
+        elif isinstance(query, (DocumentArray, Document)):
+            if isinstance(query, Document):
+                query = DocumentArray(query)
+                return self.search(query, **kwargs)[0].matches
+
+            return self.search(query, **kwargs)
+
+        try:
+            _, _ = get_array_type(query)
+            q_mat = to_numpy_array(query)
+            query = DocumentArray([Document(embedding=x) for x in q_mat])
+            if q_mat.ndim == 1:
+                query = DocumentArray(Document(embedding=q_mat))
+                return self.search(query, **kwargs)[0].matches
```

Comment:
I think it's better not to create Documents when searching by vectors. Maybe we can keep this section custom for pqlite and weaviate so that we just search by vectors, without having any documents created

Reply:
> Maybe we can keep this section custom for pqlite and weaviate so that we just search by vectors

By this way, the implementation of `search` must **validate the shape of query** to determin how to return the results. This is where error usually happens. So, in the latest commit, the signature of function `search()` looks like:
```
def search(
        self: 'T',
        query: 'DocumentArray',
        ...
) -> List['DocumentArray']
```
The search function always accepts a da query and returns a list of matched da .

Reply:
> without having any documents created

I made some revisions that the new-created docs will only be used as auxiliary, and will not being included in the result.